Performance improvement of unit tests: Optimization of waitUntil

staskus opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the GitHub issue this PR addresses. -->
Part of: #11228
Context: pdfdoF-64g-p2#comment-7238

I picked one-reviewer per team for awareness.

## Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

WooCommerce-iOS has multiple slow unit tests. [See Buildkite for slowest tests within the last 7 days](https://buildkite.com/organizations/automattic/analytics/suites/woocommerce-ios/tests?branch=trunk&filter=duration&period=7days). 

Unit tests run duration is around 7min 45sec with an average test execution duration of 85ms. Ideally, we should aim for 10ms average test duration.

There are multiple reasons why (some) unit tests are slow. Some of which are understandable and could only be deleted or moved away while others could be improved.

Examples:
- Test target includes performance tests (e.g `testViewModelGenerationTimeWithFromBigSetOfCategories`) which is slow by design
- CoreData-related tests, such as `MigrationTests`
- ViewController tests, such as `MainTabBarControllerTests` which an under-the-hood rely on UIKit
- Asynchronous tests that inefficiently wait for expectations
- Asynchronous tests that wait for something not to happen (`isInverted = true`) which requires the test to wait up until `timeout` to verify
- ....

This PR is likely the first of at least a couple of PRs. In my investigation, I try to go through the slowest tests and find common issues. I'm looking for a few key improvements that could apply to many tests. 

## Solution

The following changes improve unit test target speed from around 7:45min to 4:30min

### Improve `waitUntil`  to use a more efficient mechanism

Many of the slowest tests are relying on `waitUntil` defined in `TestKit` that waits for the condition to be true up until the timeout. 

It uses [expectation(for: predicate)](https://developer.apple.com/documentation/xctest/xctestcase/1500569-expectationforpredicate) which is convenience method for [XCTNSPredicateExpectation](https://developer.apple.com/documentation/xctest/xctnspredicateexpectation). 

As explained [in Gio's blog post](https://mokacoding.com/blog/xctnspredicateexpectation-slow/):
> XCTNSPredicateExpectation requires a timeout of at least 1.1 seconds or it will fail regardless of whether the behavior under test occurred. To avoid slowing down your tests, use [Nimble](https://github.com/Quick/Nimble)'s toEventually instead.
>
> Do yourself, your team, and your CI a favor and rewrite all your tests using XCTNSPredicateExpectation with Nimble's toEventually.

I tested replacing the `expectation(for: predicate)` with Nible's `expect(condition:).toEventually(beTrue))` and the performance improved substantially:
- Improved `waitUntil`
- Added `await until` to be used in async contexts
- Updated async test usage where necessary

Although this is an extra dependency, it only affects a test target and produces a great performance improvement:

🐢  **[Trunk](https://buildkite.com/organizations/automattic/analytics/suites/woocommerce-ios/runs/2a27fc16-c859-85f6-80ee-448043898acd)**

<img width="700" alt="image" src="https://github.com/user-attachments/assets/c23513a8-516a-418a-96ae-b497ffa36917" />

🐈  **[This Branch](https://buildkite.com/organizations/automattic/analytics/suites/woocommerce-ios/runs/29351208-5782-816f-b032-7d840a350a4d)**

<img width="700" alt="image" src="https://github.com/user-attachments/assets/65cd2df4-cbf6-4b3d-917e-0acfa5e89d30" />

### Turn asynchronous tests into synchronous ones

Along the way, I noticed some slow async tests that could be turned into synchronous ones and run in a few ms instead of a couple of seconds:

- `OrderListViewModelTests  test_banner_should_not_be_shown_when_there_is_no_error` was one of the slower tests due to a few `await` calls. However, the same logic can be tested without them. 
- `TooltipPresenterTests` would run slow because `UIView.animate` was used within tested methods. I injected a mock animation method to make the test synchronous.

## Steps to reproduce
<!-- Step-by-step testing instructions. For new user flows, consider instead stating the goal of the workflow and see if your PR reviewer can accomplish the workflow without specific steps! -->

CI should succeed.

## Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

There's a danger of introducing extra flakiness. However, after finalizing changes I ran a test suite on my MacBook and on CI dozens of times.

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [x] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [x] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [x] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

staskus committed:
Make slow OrderListViewModel tests synchronous

staskus committed:
Use Nimble toEventually to replace slow expectation(for: predicate)

expectation(for: predicate) is a wrapper for XCTNSPredicateExpectation which uses polling as opposed to more direct callback mechanisms and with a long sampling interval.

dangermattic commented:
<!--
  0 Errors
  1 Warning: <code>TestKit/Package.swift</c...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>TestKit/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages as appropriate to your project setup (e.g. in Xcode or by running <code>swift package resolve</code>).</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Add await until to be used in async context

staskus committed:
Update MainTabBarController to work with quicker waitUntil assertions

staskus committed:
Make Tooltip presenter tests synchronous by injecting UIView animation mock

10-100x speed improvement

staskus committed:
Merge branch 'trunk' into fix/11228-performance-improvement-of-slow-tests

staskus committed:
Update TestKit/Package.resolved

staskus committed:
Improvements to ViewController-containing tests

Introduction of improvement waitUntil performance makes some assertions made previously false, since view controllers are still not in the expected state when some cases of waitUntil pass.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is assigned to the mil...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is assigned to the milestone <a href="https://github.com/woocommerce/woocommerce-ios/milestone/234">21.4</a>. This milestone is due in less than 2 days.<br />
Please make sure to get it merged by then or assign it to a milestone with a later deadline.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus committed:
Merge branch 'trunk' into fix/11228-performance-improvement-of-slow-tests

staskus commented on WooCommerce/WooCommerceTests/ViewRelated/Products/Edit Product/ProductFormViewController+ProductImageUploaderTests.swift:
```diff
@@ -79,23 +79,42 @@ final class ProductFormViewController_ProductImageUploaderTests: XCTestCase {
     }
 }
+
+// MARK: - MockNavigationController
```
This is the only place in tests I found `popViewController` to be called.

iamgabrielma commented on TestKit/Sources/TestKit/XCTestCase+Wait.swift:
````diff
@@ -36,23 +37,37 @@ extension XCTestCase {
     /// ```
     ///
-    public func waitUntil(file: StaticString = #file,
+    @available(*, noasync, message: "Use await until(expression:) instead!")
````
Nice touch for awareness!

iamgabrielma reviewed (approved):
Thanks for working on this Povilas! Looks great 🚢  

As you mention it adds a bit to the dependency graph, but in exchange we're halving the target speed so IMHO it's worth the trade-off 👍
